### PR TITLE
Add unit tests for sharing map that check that views are not invalidated by modifications

### PR DESCRIPTION
**Pull request by danpoe**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: deb6390892811c504d81ba4fafe88895b78a9c3c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108723612

File: unit/util/sharing_map.cpp [hunk 1]
```diff
@@ -319,7 +343,7 @@ TEST_CASE("Sharing map views and iteration", "[core][util]")
       pairs.clear();
       for(auto &p : view)
       {
-        pairs.push_back({p.first, p.second});
+        pairs.push_back({as_string(p.first), p.second});
```
Comment (peterschrammel) [inline unit/util/sharing_map.cpp]: `id2string`

File: unit/util/sharing_map.cpp [hunk 2]
```diff
@@ -354,11 +378,11 @@ TEST_CASE("Sharing map views and iteration", "[core][util]")
     sharing_map_standardt sm;
     fill(sm);
 
-    typedef std::pair<dstringt, std::string> pt;
+    typedef std::pair<std::string, std::string> pt;
     std::vector<pt> pairs;
 
     sm.iterate([&pairs](const irep_idt &key, const std::string &value) {
-      pairs.push_back({key, value});
+      pairs.push_back({as_string(key), value});
```
Comment (peterschrammel) [inline unit/util/sharing_map.cpp]: `id2string`

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: deb6390892811c504d81ba4fafe88895b78a9c3c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108790909

File: unit/util/sharing_map.cpp [hunk 3]
```diff
@@ -156,6 +156,10 @@ TEST_CASE("Sharing map interface", "[core][util]")
     REQUIRE(sm.has_key("i"));
     REQUIRE(sm.has_key("j"));
     REQUIRE(!sm.has_key("k"));
+
+    cbmc_invariants_should_throwt invariants_throw;
+
+    REQUIRE_THROWS(sm.insert("i", "4"));
```
Comment (tautschnig) [inline unit/util/sharing_map.cpp]: I would suggest to also require a particular exception type (using `REQUIRE_THROWS_AS`). See other unit tests for examples.

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 2b4f0aed2d6c904c7225574d1d72ade5f95b2816).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108855879
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Pull request merged**